barretenberg/cpp/src/barretenberg/world_state/world_state.cpp

```diff
@@ -0,0 +1,278 @@
+#include "barretenberg/world_state/world_state.hpp"
+#include "barretenberg/crypto/merkle_tree/append_only_tree/append_only_tree.hpp"
+#include "barretenberg/crypto/merkle_tree/fixtures.hpp"
+#include "barretenberg/crypto/merkle_tree/hash_path.hpp"
+#include "barretenberg/crypto/merkle_tree/indexed_tree/indexed_leaf.hpp"
+#include "barretenberg/crypto/merkle_tree/lmdb_store/functions.hpp"
+#include "barretenberg/crypto/merkle_tree/response.hpp"
+#include "barretenberg/crypto/merkle_tree/signal.hpp"
+#include "barretenberg/world_state/history.hpp"
+#include "barretenberg/world_state/tree_with_store.hpp"
+#include <memory>
+#include <stdexcept>
+#include <tuple>
+#include <unordered_map>
+#include <utility>
+#include <variant>
+
+namespace bb::world_state {
+
+const uint WORLD_STATE_MAX_DB_COUNT = 16;
+
+using namespace bb::crypto::merkle_tree;
+
+WorldState::WorldState(uint threads, const std::string& data_dir, uint map_size_kb)
+    : _workers(threads)
+{
+    _lmdb_env = std::make_unique<LMDBEnvironment>(data_dir, map_size_kb, WORLD_STATE_MAX_DB_COUNT, threads);
+
+    {
+        const auto* name = "nullifier_tree";
+        auto lmdb_store = std::make_unique<LMDBStore>(*_lmdb_env, name, false, false, IntegerKeyCmp);
+        auto store = std::make_unique<NullifierStore>(name, NULLIFIER_TREE_HEIGHT, *lmdb_store);
+        auto tree = std::make_unique<NullifierTree>(*store, _workers, 128);
+        _trees.insert(
+            { MerkleTreeId::NULLIFIER_TREE, TreeWithStore(std::move(tree), std::move(store), std::move(lmdb_store)) });
```

Comment:
The ownership model here is very very error prone.

* `lmdb_store` has a reference to `lmdb_env` of unknown lifetime.
* `store` holds a reference to `lmdb_store` which _in principle_ will only live inside this block. However, then the lifetime gets "extended" by moving it into _trees (which hopefully lives longer than everything else).
* Same with the tree.

Some alternatives:
* Let's assume we are ok with LMDBStore having a reference to a long lived environment. This is not too strange and you know it lives as long as the instance of this class.
* AFAICT from the blocks here, you can actually transfer ownership of most of the ptrs to the next thing you are constructing:

```
auto lmdb_store = std::make_unique<LMDBStore>(*_lmdb_env, name, false, false, IntegerKeyCmp);
auto store = std::make_unique<NullifierStore>(name, NULLIFIER_TREE_HEIGHT, std::move(lmdb_store));
auto tree = std::make_unique<NullifierTree>(std::move(store), _workers, 128);
```

Then you JUST insert the tree in _trees. Of course you'd have to change the stores to take and store an std::unique_ptr instead of a reference. This is the right thing to do in this case. Ownership idioms are quite important in C++ and are usually reflected by whether you use a ptr, reference, or unique_ptr. Taking a unique_ptr usually means ownership transfer. Taking a ref/ptr usually means not. 

I think this suggestion would work better and be less error prone (you cannot forget to extend the lifetimes). But of course, if you do need somehow raw access to store and lmdb_store in this class (like you have now in _trees) then you might need to change a few things.

PS: I do like that you are using unique_ptr vs shared_ptr ;)

Reply:
Thanks @fcarreiro. I'll re-work all of the object relationships over the course of the next few PRs that we will be creating for this native world state.